### PR TITLE
Protocol Features, coverage, types, BTC markets websocket

### PR DESCRIPTION
**Pull request by gloriousCode**
image

# Description
This is a small improvement PR which fleshes out protocol features for REST and websocket. It moves currency package types into their own type files. It improves test coverage in currency and config packages. It adds BTC Markets websocket support and it also removes "test failed" strings from test error messages, because the test failing is an easy way to determine that the test has failed 🎉 

This PR is a result of doing QA on engine branch. Its being created now because its already at 5k line changes across almost 150 files. I'll be continuing improvements after this as well

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
`go test ./... -race` `golanglint-ci run` && running the application

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool 
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/websocket/wshandler/wshandler_types.go [hunk 1]
```diff
@@ -204,3 +168,42 @@ type WebsocketConnection struct {
 	ResponseMaxLimit     time.Duration
 	TrafficTimeout       time.Duration
 }
+
+// ProtocolFeatures holds all variables for the exchanges supported features
+// for a protocol
+type ProtocolFeatures struct {
```
Comment (thrasher-) [inline exchanges/websocket/wshandler/wshandler_types.go]: This introduces many copies of these big protocol structs. Would be nice to export features to its own package under exchanges so we can rid this, plus won't experience any import cycles similar to how we do it with the asset package. Also has the added benefit of not having to update them all individually.

Comment (gloriousCode) [reply exchanges/websocket/wshandler/wshandler_types.go]: Yeah, I wasn't happy with this one, but had import cycle problems. Didn't like the 3rd duplication, however golang is strict when it comes to applying it. If someone ever changed anything on one struct but not the other, it won't compile, so it wouldn't crate too much work to fix.
I'll look into moving it to its own area that doesn't create an import problem 👍

File: exchanges/localbitcoins/localbitcoins_wrapper.go [hunk 2]
```diff
@@ -9,6 +9,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/thrasher-corp/gocryptotrader/exchanges/protocol"
+
```
Comment (shazbert) [inline exchanges/localbitcoins/localbitcoins_wrapper.go]: rm line

File: exchanges/websocket/wshandler/wshandler_test.go [hunk 3]
```diff
@@ -85,6 +87,42 @@ func TestConnectionMessageErrors(t *testing.T) {
 	ws.DataHandler = make(chan interface{})
 	ws.ShutdownC = make(chan struct{})
 	ws.connector = func() error { return nil }
+	ws.features = &protocol.Features{
```
Comment (shazbert) [inline exchanges/websocket/wshandler/wshandler_test.go]: could just &protocol.Features{}

**Comment by shazbert** [comment]
image

Tested the orderbook stream through websocket; but we are out by a hundredth (.01) compared to btcmarkets front end. :man_shrugging:

File: currency/code_test.go [hunk 4]
```diff
@@ -220,50 +243,101 @@ func TestBaseCode(t *testing.T) {
...
+		t.Error("BaseCode UpdateContract error", err)
+	}
+	if main.Items[0].ID != 1338 {
+		t.Error("ID not updated")
+	}
+
+	main.Items[0].Role = Token
+	err = main.UpdateCryptocurrency("MEWOW", "CATS", 3)
+	if err == nil {
+		t.Error("Expecting cryptocurrency to already exist", err)
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply currency/code_test.go]: I'll check the whole codebase for this. I've probably done it a couple of times

**Pull request merged**